### DISCUSSION TITLE
delay next/Link

### DISCUSSION THREAD

**Discussion by onopko81** [vercel/next.js#15159] [Help]
# Feature request

How add a Route link a delay for page transitions?

**Comment by mulholo** [comment]
To delay, you could use the [router api ](https://nextjs.org/docs/api-reference/next/router)and put that inside [`setTimeout`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout).

You should consider the accessibility considerations of this, however. Links should probably take the form of `<a />` tags (or `<Link />`).

**Reply by nagman** [reply]
And how do you put the page mounting/unmounting inside a setTimeout through the router API?
You can listen to events that are triggered on link's click and on page change, but they don't allow you to delay the native React replace of the pages components

**Comment by franz-fletcher** [comment]
Would be nice to have this configurable at a higher level - like allowing [`router.beforePopState`](https://nextjs.org/docs/api-reference/next/router#routerbeforepopstate) to either expose a `next()` function as a parameter or accept promises as the `cb()`.

Just doesn't seem developer friendly to have to rebuild the wheel when page transitions are a pretty common UI pattern.

**Reply by franz-fletcher** [reply]
Whilst the above would still be nice, I managed to find success by adapting [this working example](https://codesandbox.io/s/react-spring-699-57yi5?file=/pages/_app.js) which leverages [the useTransition rework](https://aleclarson.github.io/react-spring/v9/#The-useTransition-rework) introduced in [`react-spring v9`](https://aleclarson.github.io/react-spring/v9/).

I know Framer Motion is popular for page transitions and I'm sure there is a similar solution.

I'm not a fan of having all that transition logic in my `_app.jsx` source code so I've modularised the example using [HOCs](https://reactjs.org/docs/higher-order-components.html) and [render props](https://reactjs.org/docs/render-props.html) which has afforded flexibility to implement child transitions that can be attached to any component and animate before and after page transitions. Happy to create an example and share for anyone interested.

**Comment by KenDzz** [comment]
This is the solution I used. Hope it can help

App.js

```
 const router = useRouter();
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    router.events.on("routeChangeError", (e) => setLoading(false));
    router.events.on("routeChangeStart", (e) => setLoading(false));
    router.events.on("routeChangeComplete", (e) => setTimeout(() => {setLoading(true)}, 1000));

    return () => {
      router.events.off("routeChangeError", (e) => setLoading(false));
      router.events.off("routeChangeStart", (e) => setLoading(false));
      router.events.off("routeChangeComplete", (e) => setTimeout(() => {setLoading(true)}, 1000));
    };
  }, [router.events]);
```


Next step

`{loading ? (<Component {...pageProps} />) : (null)}
`

**Comment by sendy34** [comment]
I was able to wrap the Page into a Memo, and delaying the Memo update after Route is changed and transition is complete.
This approach is better than caching the whole state object etc.

```
const Child = ({ children, readyToReRender }) => {
  return <>{children}</>
}

const MemoizedChild = React.memo(Child, (p, n) =>
  n.readyToReRender ? false : true
)

// Usage

// do transitions in useEffect and update readyToRender state after setTimeout

return (
    <MemoizedChild readyToReRender={readyToReRender}>
       {children}
     </MemoizedChild>
)
```
